### PR TITLE
Streaming variational inference: out-of-core DataLoader for minibatch ADVI

### PR DESCRIPTION
**Pull request by YichengYang-Ethan**
Draft for mentor review (GSoC 2026, Streaming Variational Inference). Not for merge yet.

Split out per review: this PR is now the out-of-core data layer only; the `Trainer` follows in a stacked PR once this lands.

Adds `pymc.variational.streaming`, an out-of-core minibatch VI path for data that does not fit in RAM. The API matches `torch.utils.data`:

- `IterableDataset` / `parquet_source`: a re-iterable, out-of-core source of rows. `parquet_source` reads one row group at a time, with the column order frozen at construction so a shard with a permuted schema cannot silently swap features.
- `DataLoader`: batches and optionally shuffles the source. Blocks of any size are re-batched in order, and a raw array works directly: `DataLoader(np.random.normal(size=(10_000, 2)), batch_size=64)` from the VI overview sketch yields `(64, 2)` row batches (`sample_shape` defaults to `arr.shape[1:]` for an array source). Trailing rows that do not fill a final batch are dropped, like `drop_last=True` in torch, because the model observes a fixed-shape placeholder; with `shuffle=True` the dropped remainder is re-drawn each epoch, while a fixed replay order keeps dropping the same tail rows (documented, and the second open question below). The loader is sized, so `total_size=len(loader)` gives the `N / batch_size` ELBO rescaling (the same mechanism as `pm.Minibatch`).

With bounded source chunks the full dataset never sits in RAM at once: the model observes only a `(batch_size, *sample_shape)` placeholder that is overwritten with the next minibatch each step, and peak memory is bounded by the batch, the source chunks in flight, and the optional shuffle buffer (concatenating a fill briefly holds a second copy) — independent of `N`.

Status: 55 tests in `tests/variational/test_streaming.py` and `test_streaming_autosize.py`. The `total_size` log-likelihood rescaling is checked numerically (scaled by exactly `N / batch_size`, the same mechanism as `pm.Minibatch`), alongside the re-batching, shuffle-buffer, `parquet_source`, and `total_size="auto"` behavior.

Open questions for discussion:
- Whether `total_size` can leave the model body (currently `len(loader)`).
- Whether the loader should carry the end-of-pass remainder into the next epoch instead of dropping it, so every row participates even without shuffling.
- `len(loader)` returns the row count `N`, matching the VI overview note (`__len__` is `total_size`), not the batch count that torch returns.

Example notebook: pymc-devs/pymc-examples#888

### CODE REVIEW

File: tests/variational/test_streaming.py [hunk 1]
```diff
@@ -0,0 +1,442 @@
...
+    assert ds.batches_seen == 0
+    ds.advance()
+    assert ds.as_tensor().get_value().shape == (4, 2)
+    assert ds.batches_seen == 1 and ds.rows_streamed == 4
+    ds.advance()
+    assert ds.batches_seen == 2 and ds.rows_streamed == 8
+
+
+def test_plain_loader_rebatches_arbitrary_blocks():
+    # blocks of 3 with batch_size=4: re-batched in order; the trailing 2 rows that
```
Comment (zaxtax) [inline tests/variational/test_streaming.py]: Reserve comments for non-obvious code flow

File: tests/variational/test_streaming.py [hunk 2]
```diff
@@ -0,0 +1,442 @@
...
+
+def test_shuffle_buffer_conserves_rows_non_dividing():
+    # buffer_size and chunk size deliberately do NOT divide batch_size: the
+    # carry-over must not lose or duplicate any row (regression for the drop bug).
+    data = np.arange(140, dtype="float64").reshape(140, 1)
+    src = shuffle_buffer(_chunks(data, 7), buffer_size=55, batch_size=10, seed=0)
+    batches = list(src())
+    assert all(b.shape == (10, 1) for b in batches)
+    seen = np.sort(np.concatenate([b.ravel() for b in batches]))
+    # 140 rows, batch 10 -> 14 full batches, nothing dropped (140 % 10 == 0)
```
Comment (zaxtax) [inline tests/variational/test_streaming.py]: Remove this comment

File: tests/variational/test_streaming.py [hunk 3]
```diff
@@ -0,0 +1,442 @@
...
+        batch_size=4,
+        sample_shape=(1,),
+        total_size=8,
+        preprocess_fn=lambda b: b * 3.0,
+    )
+    ds.advance()
+    np.testing.assert_array_equal(ds.as_tensor().get_value(), np.full((4, 1), 3.0))
+
+
+def test_shuffle_buffer_conserves_rows_non_dividing():
```
Comment (zaxtax) [inline tests/variational/test_streaming.py]: Use docstrings not comments for describing the test

File: tests/variational/test_streaming.py [hunk 4]
```diff
@@ -0,0 +1,442 @@
...
+    seen = []
+    for _ in range(12):  # 120 / 10
+        ds.advance()
+        seen.append(ds.as_tensor().get_value().copy())
+    assert all(b.shape == (10, 1) for b in seen)
+    np.testing.assert_array_equal(np.sort(np.concatenate([b.ravel() for b in seen])), data.ravel())
+
+
+def test_total_size_rescales_logp_like_minibatch():
+    # observed=buf[:, k] + total_size=N must scale the observed log-likelihood by
```
Comment (zaxtax) [inline tests/variational/test_streaming.py]: Unclear comment

File: tests/variational/test_streaming.py [hunk 5]
```diff
@@ -0,0 +1,442 @@
...
+    ds = DataLoader(
+        _chunks(data, 4), batch_size=np.int64(4), sample_shape=(1,), total_size=np.int64(8)
+    )
+    ds.advance()
+    assert ds.batch_size == 4
+    with pytest.raises(ValueError):
+        DataLoader(_chunks(data, 4), batch_size=True, sample_shape=(1,), total_size=8)
+
+
+def test_shuffle_buffer_reshuffles_across_epochs():
```
Comment (zaxtax) [inline tests/variational/test_streaming.py]: This comment would be better done with a more descriptive test name

File: pymc/variational/streaming.py [hunk 6]
```diff
@@ -0,0 +1,811 @@
...
+        self._warned_size = False  # the sanity check below fires at most once
+
+        # The persistent stream behind advance()/as_tensor(); re-batched like __iter__.
+        self._source_iter: Iterator[np.ndarray] = self._rebatched()
+
+        self._shared = pytensor.shared(
+            np.zeros((batch_size, *self._sample_shape), dtype=dtype), name=name
+        )
+
+    # ----- read-only state ---------------------------------------------------
```
Comment (zaxtax) [inline pymc/variational/streaming.py]: Remove all these banners

File: pymc/variational/streaming.py [hunk 7]
```diff
@@ -0,0 +1,811 @@
...
+    def as_tensor(self) -> pt.TensorVariable:
+        """A ``pytensor.shared`` buffer the model can observe directly.
+
+        An alternative to the recommended ``pm.Data`` + :class:`Trainer` path: drive
+        ``pm.fit`` yourself and call :meth:`advance` (e.g. from a callback) to refill
+        this buffer each step.
+        """
+        return self._shared
+
+    def advance(self) -> None:
```
Comment (zaxtax) [inline pymc/variational/streaming.py]: Is `advance`  used in any public API here or in pytorch?

File: pymc/variational/streaming.py [hunk 8]
```diff
@@ -0,0 +1,811 @@
...
+    data_name : str, default "batch"
+        Name of the ``pm.Data`` placeholder minibatches are streamed into. Must
+        match the name used for ``pm.Data(name, ...)`` in the model.
+    **fit_kwargs
+        Default keyword arguments forwarded to :func:`pymc.fit` (e.g.
+        ``obj_optimizer``); per-call kwargs to :meth:`fit` override them.
+
+    Notes
+    -----
+    This is the *starting point* Rob suggested: the per-step ``set_data`` logic
```
Comment (zaxtax) [inline pymc/variational/streaming.py]: I hope this note won't be here once it's draft ready ;)

File: pymc/variational/streaming.py [hunk 9]
```diff
@@ -0,0 +1,811 @@
...
+        *,
+        random_seed: int | None = None,
+        progressbar: bool = False,
+        **kwargs,
+    ):
+        """Fit for ``n`` steps, streaming minibatches into the model's placeholder.
+
+        Returns whatever :func:`pymc.fit` returns for the chosen method.
+        """
+        from pymc.model import modelcontext
```
Comment (zaxtax) [inline pymc/variational/streaming.py]: I would do import as instead of an internal import here.

File: tests/variational/test_streaming.py [hunk 10]
```diff
@@ -0,0 +1,442 @@
...
+def test_shuffle_buffer_does_not_mutate_source():
+    data = np.arange(100, dtype="float64").reshape(100, 1)
+    original = data.copy()
+    src = shuffle_buffer(_chunks(data, 25), buffer_size=40, batch_size=10, seed=1)
+    list(src())
+    np.testing.assert_array_equal(data, original)  # source untouched
+
+
+def test_dataloader_shuffle_true_yields_full_batches():
+    # shuffle=True wraps the source in a bounded shuffle_buffer internally; batches
```
Comment (zaxtax) [inline tests/variational/test_streaming.py]: Should be a docstring not a comment. Please apply this to the rest of the tests

**Review by zaxtax — COMMENTED** [review]
Leaving some comments as you prepare the draft!

File: tests/variational/test_streaming.py (hunk 1 above)
Comment (YichengYang-Ethan) [reply tests/variational/test_streaming.py]: Done, every test describes itself in a docstring now

File: pymc/variational/streaming.py (hunk 6 above)
Comment (YichengYang-Ethan) [reply pymc/variational/streaming.py]: Done

File: pymc/variational/streaming.py (hunk 7 above)
Comment (YichengYang-Ethan) [reply pymc/variational/streaming.py]: No,  torch's DataLoader has no `advance` and there's no public one here; I removed that path the only `advance` left is a private `_advance` closure inside `Trainer.fit` that drives `set_data`

File: tests/variational/test_streaming.py [hunk 11]
```diff
@@ -0,0 +1,619 @@
...
+
+from pymc.variational.streaming import (
+    DataLoader,
+    IterableDataset,
+    Trainer,
+    shuffle_buffer,
+)
+
+
+def _chunks(data, size):
```
Comment (zaxtax) [inline tests/variational/test_streaming.py]: Why is this here?

Comment (YichengYang-Ethan) [reply tests/variational/test_streaming.py]: fresh iterator each call (a bare generator would be one-shot). It yields the data a chunk of rows at a time, the way an out-of-core source like I can fold into a conftest if you'd prefer.

**Comment by fonnesbeck** [comment]
@ricardoV94 Do we want this here or in extras?

**Comment by ricardoV94** [comment]
If it can go in extras, let's go there first

**Comment by zaxtax** [comment]
@YichengYang-Ethan can you rebase this against main and mark it ready for review?

**Comment by YichengYang-Ethan** [comment]
> @YichengYang-Ethan can you rebase this against main and mark it ready for review?

Done, rebased against main and marked it ready.

**Comment by ricardoV94** [comment]
isn't this going to extras?

**Comment by zaxtax** [comment]
oh wait yes

**Comment by YichengYang-Ethan** [comment]
Ported to pymc-extras as #698 following the extras-first discussion — closing the core version to keep review in one place. Thanks!
